python/examples/manual_message_decode.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python3
+
+import logging
+import os
+import sys
+
+# Add the Python root directory (fusion-engine-client/python/) to the import search path to enable FusionEngine imports
+# if this application is being run directly out of the repository and is not installed as a pip package.
+root_dir = os.path.normpath(os.path.join(os.path.dirname(__file__), '..'))
+sys.path.insert(0, root_dir)
+
+from fusion_engine_client.messages.core import *
+from fusion_engine_client.messages import ros
+from fusion_engine_client.utils.argument_parser import ArgumentParser
+from fusion_engine_client.utils.log import locate_log, DEFAULT_LOG_BASE_DIR
+
+from examples.message_decode import print_message
+
+
+logger = logging.getLogger('point_one.fusion_engine')
+
+
+def decode_message(header, data, offset):
+    # Validate the message length and CRC.
+    if len(data) != header.calcsize() + header.payload_size_bytes:
+        return False
+    else:
+        header.validate_crc(data)
+
+    # Check that the sequence number increments as expected.
+    if header.sequence_number != decode_message.expected_sequence_number:
+        logger.info('Warning: unexpected sequence number. [expected=%d, received=%d]' %
+                    (decode_message.expected_sequence_number, header.sequence_number))
+
+    decode_message.expected_sequence_number = header.sequence_number + 1
+
+    # Deserialize and print the message contents.
+    #
+    # Note: This could also be done more generally using the fusion_engine_client.message_type_to_class dictionary.
+    # We do it explicitly here for sake of example.
+    if header.message_type == PoseMessage.MESSAGE_TYPE:
+        contents = PoseMessage()
+        contents.unpack(buffer=data, offset=offset)
+    elif header.message_type == GNSSInfoMessage.MESSAGE_TYPE:
+        contents = GNSSInfoMessage()
+        contents.unpack(buffer=data, offset=offset)
+    elif header.message_type == GNSSSatelliteMessage.MESSAGE_TYPE:
+        contents = GNSSSatelliteMessage()
+        contents.unpack(buffer=data, offset=offset)
+    elif header.message_type == ros.ROSPoseMessage.MESSAGE_TYPE:
+        contents = ros.ROSPoseMessage()
+        contents.unpack(buffer=data, offset=offset)
+    elif header.message_type == ros.ROSGPSFixMessage.MESSAGE_TYPE:
+        contents = ros.ROSGPSFixMessage()
+        contents.unpack(buffer=data, offset=offset)
+    elif header.message_type == ros.ROSIMUMessage.MESSAGE_TYPE:
+        contents = ros.ROSIMUMessage()
+        contents.unpack(buffer=data, offset=offset)
+    else:
+        contents = None
+
+    print_message(header, contents)
+
+    return True
+
+
+decode_message.expected_sequence_number = 0
```

Comment:
Hunh, haven't seen that before. What exactly is `decode_message` referencing here? Is it adding an attribute to the function?

Reply:
Yep, it's a function attribute. Basically the same as just making this a global, except without the dumb syntax limitation that you can't set a global primitive inside a function so you always have to make the global a list or something.